File: Leetcode_98.py
#Time Complexity: O(n) 
#Space Complexity: O(n)

class  TreeNode(object):
    def _init_(self,x):
        self.val=x
        self.left=None
        self.right= None
class Solution:
    def isValidBST(self, root):
        self.correct=True
        self.prev=float('-inf')
        self.inorder(root)
        return self.correct
    def inorder(self, node):
        if not node or not self.correct:
            return
        self.inorder(node.left)
        if  node.val<=self.prev:
            self.correct=False
            return
        self.prev=node.val
        self.inorder(node.right)
        
class Solution2(object):
    def isValidBST(self,root):
        return self.valid(root,float('inf'), float('inf'))
    def valid(self, node, lower, upper):
        if not node:
            return True
        if node.val<= lower or node.val>=upper:
            return False
        return self.valid(node.left, lower, node.val) and self.valid(node.right, node.val, upper)

File: Leetcode98.java
import java.util.Stack;

public class Leetcode98 {

    class TreeNode {
            int val;
            TreeNode left;
            TreeNode right;

            TreeNode() {
            }

            TreeNode(int val) {
                this.val = val;
            }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    public boolean isValidBST(TreeNode root) {
        // this pre order traversal, and in order to valid a binary tree is bst, if we
        // traverse the
        // the binary tree in inoder travel then we will get a incresing order and if
        // order get
        // distrub any time then its not a bst.
        // we will use iterative method of tree traversal, and with iterative method we
        // have to
        // use the stack.

        if (root == null)
            return false;

        Stack<TreeNode> st = new Stack<>();
        TreeNode prev = null;

        while (root != null || !st.isEmpty()) {
            while (root != null) {
                st.push(root);
                root = root.left;
            }

            root = st.pop();
            if (prev != null && prev.val >= root.val) {
                return false;
            }
            prev = root;
            System.out.print(root.val);
            root = root.right;
        }

        return true;
    }
}

// time complexity==O(n)
// space complexity==O(height of tree)